### DISCUSSION TITLE
Everything a Plug-in

### DISCUSSION THREAD

**Discussion by sirinath** [GraphiteEditor/Graphite#1466] [Ideas/Feedback]
Make everything a plug in, but with a final running programme fully optimised.

This way the end users can add, remove and modify Graphite as needed. Maybe you can borrow a page from Bevy Engine (https://bevyengine.org) here.

The plug-in can be for the IDE or the engine.

In this case adding or changing plug-ins should be very simple. Any user should be able to do it. Also there should be protection for why buggy plugins, where partial processing data can be cached and plug-in updates can be applied to certain parts to minimise render time. Also processing will be needed if the input changes or the pipeline / graph changes. You can borrow a page from Comemo (https://github.com/typst/comemo).

The user need not worry about building and compiling the plugins. This can be transparently done, when configuration changes. Certain plugins can be loaded from configured code folders.

This will also help quickly prototype new versions of Graphite and and open up to direct contribution Graphite itself or the ecosystem from any average user. You don't need to be a proper / professional developer or programmer to contribute or to hack Graphite. You will need some coding skill thought.

**Comment by Keavon** [comment]
This remains yet to be figured out but we'll generally lean towards most extensibility occurring through the power of the node graph with a shareable, reusable custom nodes ecosystem.
